garbage-collection/gc_cleanup.py

```diff
@@ -0,0 +1,148 @@
+def clean_up(args):
+    # body object for kubernetes api
+    body = client.V1DeleteOptions()
+    # get all workflows
+    try:
+        workflows = custom_api.list_namespaced_custom_object(args.group, args.version, args.namespace, args.plural)
+    except ApiException as e:
+        logging.info("Exception when calling CustomObjectsApi->list_namespaced_custom_object: %s\n" % e)
+
+    # track workflows expired, workflows not expired and pods deleted for logging
+    workflows_expired = []
+    workflows_not_expired = []
+    pods_deleted = []
+    for workflow in workflows['items']:
+        key = workflow['metadata']['name']
+
+        try:
+            finished_at = datetime.strptime(workflow['status']['finishedAt'], '%Y-%m-%dT%H:%M:%SZ')
+        except TypeError:
+            logging.info('could not read workflow {}'.format(key))
+        time_since_completion = (datetime.utcnow() - finished_at).total_seconds()/60/60
+        # Get specific metadata based on workflow type
+        if args.adhoc:
+            exists = False
+            for word in args.starts_with + args.label_selector:
+                if key.startswith(word):
+                    #skip
+                    exists = True
+                elif word in workflow['metadata']['labels']:
+                    #skip
+                    exists = True
+            if not exists and int(time_since_completion) > int(args.max_age_hrs):
+                workflows_expired.append(key)
+                # print("deleting workflow pod {} that completed {} hrs ago".format(key,time_since_completion))
+                pods = get_pods(workflow,args.namespace)
+                for pod in pods:
+                    if not args.dry_run:
+                        delete_pods(pod,args.namespace,body)
+                    else:
+                        logging.info("dry_run flag set, would have deleted {}".format(pod))
+            else:
+                workflows_not_expired.append(key)
+        else:
+            exists = False
+            for word in args.starts_with + args.label_selector:
+                if key.startswith(word):
+                    #skip
+                    exists = True
+                elif word in workflow['metadata']['labels']:
+                    #skip
+                    exists = True
+            if exists and int(time_since_completion) > int(args.max_age_hrs):
+                workflows_expired.append(key)
+                # print("deleting workflow pod {} that completed {} hrs ago".format(key,time_since_completion))
+                pods = get_pods(workflow, args.namespace)
+                for pod in pods:
+                    if not args.dry_run:
+                        delete_pods(pod,args.namespace,body)
+                    else:
+                        logging.info("dry_run flag set, would have deleted {}".format(pod))
+            else:
+                workflows_not_expired.append(key)
+    logging.info("expired workflows: {}".format(workflows_expired))
```

Comment:
if we're using logging (we should!), let's replace all print statements with logging